### ISSUE TITLE
Synchronous tests will not run when any asyncio_cooperative test fail

### ISSUE THREAD

**Issue by remidebette** [willemt/pytest-asyncio-cooperative#3]
Hi, 

In link with the [previous issue](https://github.com/willemt/pytest-asyncio-cooperative/issues/1), I created sample code for the issue with synchronous fixtures.
Basically, there is still an issue with synchronous fixtures, and also any failure in an `asyncio_cooperative` test removes all (synchronous) unmarked tests from the pytest report.
(we see a !!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!! in place of the remaining synchronous tests)

The following is a first test for reference. It is successful
```python
import httpx
import pytest
from pytest import fixture


@fixture()
async def client():
    async with httpx.AsyncClient() as client:
        yield client

@pytest.mark.asyncio_cooperative
async def test_google(client):
    response = await client.get("https://google.com")

    assert response.status_code == httpx.codes.ok
```

Now, if I start creating a class, that I itself will want to use as a fixture.
```python
class MyAPI:
    def __init__(self, client):
        self.client = client

    # This call is Network bound, so a use case for async
    async def get(self, url):
        return await self.client.get(url)

    # But the class also has synchronous functions
    def cpu_task(self, *args):
        return sum(args)
```

If I use a synchronous fixture for instanciating the MyAPI class, there are 2 issues:
* the `test_google_fixture` test fails.
* as soon as an `asyncio_cooperative` test fails, pytest does not try to move along to the `test_cpu_task`. 

```python
@fixture()
def my_api(client):
    return MyAPI(client)


@pytest.mark.asyncio_cooperative
async def test_google_fixture(my_api):
    response = await my_api.get("https://google.com")

    assert response.status_code == httpx.codes.ok
    # Fail

def test_cpu_task(my_api):
    response = my_api.cpu_task(2, 3)

    assert response == 5
    # Success, but is displayed in the pytest report only if I skip the failing `asyncio_cooperative` tests
```

Here is the corresponding `test_google_fixture` error:
```python
Testing started at 22:30 ...
/home/remi/.virtualenvs/printful-httpx-iQJNiyny/bin/python /home/remi/.local/share/JetBrains/Toolbox/apps/PyCharm-P/ch-0/201.7223.92/plugins/python/helpers/pycharm/_jb_pytest_runner.py --path /home/remi/PycharmProjects/printful-httpx/tests/api -- -vv -m "not slow"
Launching pytest with arguments -vv -m not slow /home/remi/PycharmProjects/printful-httpx/tests/api in /home/remi/PycharmProjects/printful-httpx

============================= test session starts ==============================
platform linux -- Python 3.8.2, pytest-5.4.2, py-1.8.1, pluggy-0.13.1 -- /home/remi/.virtualenvs/printful-httpx-iQJNiyny/bin/python
cachedir: .pytest_cache
rootdir: /home/remi/PycharmProjects/printful-httpx
plugins: asyncio-cooperative-0.8.5, dotenv-0.4.0
collecting ... collected 4 items

tests/api/test_api.py::test_api 
tests/api/test_api.py::test_google_fixture PASSED                                   [ 25%]FAILED                        [ 50%]
tests/api/test_api.py:32 (test_google_fixture)
>   item.runtest = lambda: result.result()

../../.virtualenvs/printful-httpx-iQJNiyny/lib/python3.8/site-packages/pytest_asyncio_cooperative/plugin.py:384: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../.virtualenvs/printful-httpx-iQJNiyny/lib/python3.8/site-packages/pytest_asyncio_cooperative/plugin.py:237: in test_wrapper
    fixture_values, teardowns = await fill_fixtures(item)
../../.virtualenvs/printful-httpx-iQJNiyny/lib/python3.8/site-packages/pytest_asyncio_cooperative/plugin.py:222: in fill_fixtures
    value, teardowns2 = await fill_fixture_fixtures(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

_fixtureinfo = FuncFixtureInfo(argnames=('my_api',), initialnames=('my_api',), names_closure=['my_api', 'client'], name2fixturedefs={...nction' baseid='tests/api/test_api.py'>,), 'client': (<FixtureDef argname='client' scope='function' baseid='tests'>,)})
fixture = <FixtureDef argname='my_api' scope='function' baseid='tests/api/test_api.py'>
item = <Function test_google_fixture>

    async def fill_fixture_fixtures(_fixtureinfo, fixture, item):
    
        if inspect.isasyncgenfunction(fixture.func) or isinstance(
            fixture.func, CachedAsyncGen
        ):
            fixture_values, teardowns = await _fill_fixture_fixtures(
                _fixtureinfo, fixture, item
            )
    
            # Cache the module call
            if fixture.scope in ["module"]:
                if not isinstance(fixture.func, CachedAsyncGen):
                    fixture.func = CachedAsyncGen(fixture.func)
    
            gen = fixture.func(*fixture_values)
            value = await gen.__anext__()
            return value, teardowns + [gen]
    
        elif inspect.iscoroutinefunction(fixture.func) or isinstance(
            fixture.func, CachedFunction
        ):
            fixture_values, teardowns = await _fill_fixture_fixtures(
                _fixtureinfo, fixture, item
            )
    
            # Cache the module call
            if fixture.scope in ["module"]:
                if not isinstance(fixture.func, CachedFunction):
                    fixture.func = CachedFunction(fixture.func)
    
            value = await fixture.func(*fixture_values)
            return value, teardowns
    
        # Regular fixture
        # FIXME: we should use more of pytest's fixture system
        elif inspect.isgeneratorfunction(fixture.func):
            gen = fixture.func()
            return gen.__next__(), [gen]
    
        # It's a parameterization
        # FIXME: we should use more of pytest's fixture system
        elif fixture.params:
            request = item._request
            request.param = item._pyfuncitem.callspec.params[fixture.argname]
            val = fixture.func(request)
            return val, []
    
        # FIXME: we should use more of pytest's fixture system
        elif inspect.isfunction(fixture.func):
>           val = fixture.func()
E           TypeError: my_api() missing 1 required positional argument: 'client'

../../.virtualenvs/printful-httpx-iQJNiyny/lib/python3.8/site-packages/pytest_asyncio_cooperative/plugin.py:192: TypeError
PASSED                                [ 75%]
Assertion failed

Assertion failed

tests/api/test_api.py::test_google 

=================================== FAILURES ===================================
_____________________________ test_google_fixture ______________________________

>   item.runtest = lambda: result.result()

../../.virtualenvs/printful-httpx-iQJNiyny/lib/python3.8/site-packages/pytest_asyncio_cooperative/plugin.py:384: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../.virtualenvs/printful-httpx-iQJNiyny/lib/python3.8/site-packages/pytest_asyncio_cooperative/plugin.py:237: in test_wrapper
    fixture_values, teardowns = await fill_fixtures(item)
../../.virtualenvs/printful-httpx-iQJNiyny/lib/python3.8/site-packages/pytest_asyncio_cooperative/plugin.py:222: in fill_fixtures
    value, teardowns2 = await fill_fixture_fixtures(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

_fixtureinfo = FuncFixtureInfo(argnames=('my_api',), initialnames=('my_api',), names_closure=['my_api', 'client'], name2fixturedefs={...nction' baseid='tests/api/test_api.py'>,), 'client': (<FixtureDef argname='client' scope='function' baseid='tests'>,)})
fixture = <FixtureDef argname='my_api' scope='function' baseid='tests/api/test_api.py'>
item = <Function test_google_fixture>

    async def fill_fixture_fixtures(_fixtureinfo, fixture, item):
    
        if inspect.isasyncgenfunction(fixture.func) or isinstance(
            fixture.func, CachedAsyncGen
        ):
            fixture_values, teardowns = await _fill_fixture_fixtures(
                _fixtureinfo, fixture, item
            )
    
            # Cache the module call
            if fixture.scope in ["module"]:
                if not isinstance(fixture.func, CachedAsyncGen):
                    fixture.func = CachedAsyncGen(fixture.func)
    
            gen = fixture.func(*fixture_values)
            value = await gen.__anext__()
            return value, teardowns + [gen]
    
        elif inspect.iscoroutinefunction(fixture.func) or isinstance(
            fixture.func, CachedFunction
        ):
            fixture_values, teardowns = await _fill_fixture_fixtures(
                _fixtureinfo, fixture, item
            )
    
            # Cache the module call
            if fixture.scope in ["module"]:
                if not isinstance(fixture.func, CachedFunction):
                    fixture.func = CachedFunction(fixture.func)
    
            value = await fixture.func(*fixture_values)
            return value, teardowns
    
        # Regular fixture
        # FIXME: we should use more of pytest's fixture system
        elif inspect.isgeneratorfunction(fixture.func):
            gen = fixture.func()
            return gen.__next__(), [gen]
    
        # It's a parameterization
        # FIXME: we should use more of pytest's fixture system
        elif fixture.params:
            request = item._request
            request.param = item._pyfuncitem.callspec.params[fixture.argname]
            val = fixture.func(request)
            return val, []
    
        # FIXME: we should use more of pytest's fixture system
        elif inspect.isfunction(fixture.func):
>           val = fixture.func()
E           TypeError: my_api() missing 1 required positional argument: 'client'

../../.virtualenvs/printful-httpx-iQJNiyny/lib/python3.8/site-packages/pytest_asyncio_cooperative/plugin.py:192: TypeError
=========================== short test summary info ============================
FAILED tests/api/test_api.py::test_google_fixture - TypeError: my_api() missi...
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
========================= 1 failed, 2 passed in 0.19s ==========================

```

On the other hand, if I artificially use an asynchronous fixture for instanciating the MyAPI class, it fixes the behavior of `test_google_fixture`. But then it is the synchronous test that fails this time.
```python
@fixture()
async def my_api(client):
    return MyAPI(client)


@pytest.mark.asyncio_cooperative
async def test_google_fixture(my_api):
    response = await my_api.get("https://google.com")

    assert response.status_code == httpx.codes.ok
    # Success

def test_cpu_task(my_api):
    response = my_api.cpu_task(2, 3)

    assert response == 5
    # Fail
```

Error of `test_cpu_task`:
```python
=================================== FAILURES ===================================
________________________________ test_cpu_task _________________________________

my_api = <coroutine object my_api at 0x7f2258ea61c0>

    def test_cpu_task(my_api):
>       response = my_api.cpu_task(2, 3)
E       AttributeError: 'coroutine' object has no attribute 'cpu_task'

tests/api/test_api.py:41: AttributeError
=========================== short test summary info ============================
FAILED tests/api/test_api.py::test_cpu_task - AttributeError: 'coroutine' obj...
========================= 1 failed, 3 passed in 0.14s ==========================
sys:1: RuntimeWarning: coroutine 'my_api' was never awaited

```

**Comment by willemt** [comment]
**Synchronous tests failing to run when async tests fail**
This is a known issue. I've added details about it on the README. https://github.com/willemt/pytest-asyncio-cooperative/commit/e02200c9152d061c010c9f9ff672dc79edb88982

We need to figure out a way that the synchronous tests continue to run when async tests fail (`help wanted` label added)

Personally, this is a low priority task. If this happens the test suite correctly shows there was a failure (in the async test). You can't have false negatives if you never report the test's status 😅 

**Synchronous tests failing to use async fixtures is a pytest issue**
There are no plans to support this because it is standard pytest behaviour. asyncio-cooperative avoids manipulating tests without the `asyncio_cooperative` mark

**Comment by remidebette** [comment]
Understood,
I will raise the pytest issue to them (unless you already know of an existing github ticket for this?)

Also I understand that this issue will track "synchronous tests continue to run" and that help is needed.

I would be interested to have a look. 
I am quite new to asyncio and I do not know the pytest architecture.
Would you have some "How to contribute?" guidelines or even some interesting reads you recommend before contributing to this repo?

Thanks

**Comment by remidebette** [comment]
Also, I confirm the other point that I raised in this issue (fixtures that themselves have parent fixtures such as `test_google_fixture)` is fixed with db82fcf, and works in version 0.9.0

**Comment by SilenceHe** [comment]
Any update for this issue? Looks it's a good solution to run tests in parallel, I really want this pytest plugin now, It's will be great if this issue get fixed.

**Comment by willemt** [comment]
hi @SilenceHe, I just released version `0.14.0` which has the fix

**Comment by SilenceHe** [comment]
> hi @SilenceHe, I just released version `0.14.0` which has the fix

Great, glad to hear that!!! It‘s available now with the leatest version on PyPI , thansks for your work.
